src/service/cached-xhr-impl.js

```diff
@@ -0,0 +1,164 @@
+import {Cache} from '../utils/cache';
+import {Xhr} from './xhr-impl';
+
+/**
+ * The "init" argument of the Fetch API. Currently, only "credentials: include"
+ * is implemented.  Note ampCors with explicit false indicates that
+ * __amp_source_origin should not be appended to the URL to allow for
+ * potential caching or response across pages.
+ *
+ * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+ *
+ * @typedef {{
+ *   body: (!Object|!Array|undefined|string),
+ *   credentials: (string|undefined),
+ *   headers: (!Object|undefined),
+ *   method: (string|undefined),
+ *   requireAmpResponseSourceOrigin: (boolean|undefined),
+ *   ampCors: (boolean|undefined)
+ * }}
+ */
+let FetchInitDef;
+
+
+/** @private @const {!Array<function(*):boolean>} */
+const allowedJsonBodyTypes_ = [isArray, isObject];
+
+
+/**
+ * A service that polyfills Fetch API for use within AMP.
+ *
+ * @package Visible for type.
+ * @visibleForTesting
+ */
+export class CachedXhr extends Xhr {
+
+  /**
+   * @param {!Window} win
+   * @param {number=} cacheSize
+   */
+  constructor(win, cacheSize) {
+    super(win);
+
+    /** @const {!Cache} */
+    this.fragmentCache_ = new Cache(cacheSize);
+  }
+
+  /**
+   * Fetches and constructs JSON object based on the fetch polyfill.
+   *
+   * See https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch
+   *
+   * See `fetchAmpCors_` for more detail.
+   *
+   * @param {string} input
+   * @param {?FetchInitDef=} opt_init
+   * @return {!Promise<!JSONType>}
+   * @override
+   */
+  fetchJson(input, opt_init) {
+    const init = this.setupInit_(opt_init, 'application/json');
+    const parsedInput = parseUrl(input);
+    const propertyPath = parsedInput.hash.slice(1); // remove # prefix
+    const getResponseJson = response => response.json();
+
+    if (init.method === 'POST' && !isFormData(init.body)) {
+      // Assume JSON strict mode where only objects or arrays are allowed
+      // as body.
+      dev().assert(
+        allowedJsonBodyTypes_.some(test => test(init.body)),
+        'body must be of type object or array. %s',
+        init.body
+      );
+
+      init.headers['Content-Type'] = 'application/json;charset=utf-8';
+      init.body = JSON.stringify(init.body);
+    }
+
+    const isCacheable = (init.method === 'GET');
+    if (isCacheable) {
+      const getPropertyAtPath = getPath.bind(null, propertyPath);
+      const cacheKey = this.getCacheKey_(input, opt_init);
+      const cachedPromise = this.fragmentCache_.get(cacheKey);
+      if (cachedPromise) {
+        return cachedPromise.then(getPropertyAtPath);
+      }
+
+      const fetchPromise = this.fetch(input, init).then(getResponseJson);
+
+      // Since a fragment is present, cache the full response promise, then
+      // return a promise with just the value specified by the fragment.
+      this.fragmentCache_.put(cacheKey, fetchPromise);
+      return fetchPromise.then(getPropertyAtPath);
```

Comment:
I belive the intended use-case for the grahment is e-commerce JSON APIs that return a big user-object with nested data that are displayed on many different parts of the page. For example, a user object with a name that is displayed at the top and bottom, a shopping cart list, etc. and each piece of the UI that references a deep property of the user-object would specify which property through the fragment. Does that answer your question?